### ISSUE TITLE
torch.max return wrong result on selected tensor 

### ISSUE THREAD

**Issue by Teaonly** [pytorch/pytorch#1036]
torch.max return wrong result on selected tensor.

```
In [3]: x = torch.rand(10,10)

In [4]: y = x[:, 2]                ## this is a selected tensor

In [5]: torch.max(y, 0)
Out[5]:
(
  0.6744
 [torch.FloatTensor of size 1],
  0                                        ## this is wrong 
 [torch.LongTensor of size 1])

In [6]: y = x[:, 2].clone()     ## same data with previous y

In [7]: torch.max(y, 0)
Out[7]:
(
  0.6744
 [torch.FloatTensor of size 1],
  3
 [torch.LongTensor of size 1])

In [8]: y
Out[8]:

 0.1533
 0.4855
 0.1404
 0.6744
 0.0296
 0.2307
 0.5875
 0.1057
 0.1528
 0.2445
[torch.FloatTensor of size 10]
```

**Comment by chenyuntc** [comment]
well, it works fine in my computer. Maybe you should update pytorch to the latest version

**Comment by Teaonly** [comment]
I have try head version from master, still bug

```
In [1]: import torch

In [2]: x = torch.rand(10, 10)

In [3]: y = x[:, 3]

In [4]: y
Out[4]:

 0.0497
 0.6632
 0.7263
 0.5071
 0.7504
 0.3432
 0.7698
 0.5947
 0.6670
 0.0003
[torch.FloatTensor of size 10]

In [5]: torch.max(y, 0)
Out[5]:
(
  0.7698
 [torch.FloatTensor of size 1],
  0
 [torch.LongTensor of size 1])

In [6]: y = x[:, 3].clone()

In [7]: y
Out[7]:

 0.0497
 0.6632
 0.7263
 0.5071
 0.7504
 0.3432
 0.7698
 0.5947
 0.6670
 0.0003
[torch.FloatTensor of size 10]

In [8]: torch.max(y, 0)
Out[8]:
(
  0.7698
 [torch.FloatTensor of size 1],
  6
 [torch.LongTensor of size 1])
```

**Comment by chenyuntc** [comment]
You are right. the first time when I tried,  the max index is 0, so it both return 0.
it seems that if we use slice , the index location return by`torch.max` is always `0`,

**Comment by Teaonly** [comment]
@chenyuntc 

There is no bug in Torch7, we should not change TH lib's source. 

```

  ______             __   |  Torch7
 /_  __/__  ________/ /   |  Scientific computing for Lua.
  / / / _ \/ __/ __/ _ \  |  Type ? for help
 /_/  \___/_/  \__/_//_/  |  https://github.com/torch
                          |  http://torch.ch

th> require('torch');
                                                                      [0.0000s]
th> x = torch.rand(10, 10)
                                                                      [0.0001s]
th> y = x[{{},3}]
                                                                      [0.0001s]
th> torch.max(y,1)
 0.9540
[torch.DoubleTensor of size 1]

 7
[torch.LongTensor of size 1]

                                                                      [0.0004s]
```

**Comment by adamlerer** [comment]
Yes, there's a bug in torch. I have a fix for it, will send a PR later today. Thanks!

**Comment by adamlerer** [comment]
In case anyone's interested, here's what happened:

The TH_TENSOR_APPLYX has a fast inner loop and then an outer loop where all the tensor dimensions are updated. These use different loop counters. In `min` and `max` you have to look at the correct loop counter to know which index the max occurred at. I was always looking at the outer loop counter. TH_TENSOR_APPLYX is usually only used when the max is over the outer dim, but in this particular case you're taking the max over the dimension in the inner loop.

**Comment by soumith** [comment]
fixed in master via https://github.com/pytorch/pytorch/commit/ef183a1d23be88edaa80581fc4b2282df3ed46ca
